Bug 1756475: Add flag 'DisableRedirect', decide to route all data through the registry

steveteuber opened the pull request:
When using Ceph as `s3` backend, the registry should send requests to the Rados Gateway directly.
Otherwise parsing errors will occur:

```
FATA[0000] pulling image failed: rpc error: code = Unknown desc = Error parsing image configuration: Get https://<hostname>/<bucker>/docker/registry/v2/blobs/sha256/53/53dc2a2e3904d6b911a83b228bf0c48a4dd434fbdbff29638e4b24dfb3caaf7b/data?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=<credential>%2Fdefault%2Fs3%2Faws4_request&X-Amz-Date=20190910T134717Z&X-Amz-Expires=1200&X-Amz-SignedHeaders=host&X-Amz-Signature=<signature>: EOF
```

More information about `redirect`:
https://docs.docker.com/registry/configuration/#redirect

bparees commented on pkg/apis/imageregistry/v1/types.go:
```diff
@@ -125,6 +125,10 @@ type ImageRegistrySpec struct {
+	// Redirect controls whether to route all data through the Registry,
+	// rather than redirecting to the backend.
+	// +optional
+	Redirect bool `json:"redirect"`
```
The current behavior defaults this to true, right?  So we can't change that default behavior.  This field needs to default to enabling redirect when it is not set.  (ie rename it to DisableRedirect)

steveteuber replied:
Yes, you're right. I'll rename it to `DisableRedirect `.

adambkaplan commented:
/hold

Awaiting `master` to be open for 4.3 enhancements.

steveteuber commented:
@adambkaplan Can you merge this change and cherry-pick them to branch {{release-4.2}}?

dmage commented:
/approve

The second commit amends almost every change from the first one, it's better to squash these commits.

dmage commented:
/hold cancel

bparees commented:
> @adambkaplan Can you merge this change and cherry-pick them to branch release-4.2?

just to set expectations here, 4.2 is currently locked down as we get ready for GA, it's going to be 3-4 weeks before we are able to backport this and get it out in a 4.2.z (hopefully 4.2.1) release.  but we can certainly open the PR (and BZ paperwork) now to get it ready.

adambkaplan commented:
@steveteuber does the redirect behavior impact customer use cases/expectations for 4.2? I'm hesistant to backport a this without some level of field input. (I consider this PR an enhancement, not a bug fix).

bparees commented:
> @steveteuber does the redirect behavior impact customer use cases/expectations for 4.2? 

i think we've seen evidence that customers need this capability because their docker clients don't have direct access to s3 and thus the redirect breaks them(I've seen at least 2 reports/questions about this).  While I agree it's fundamentally an enhancement, i think it's small enough and low risk enough that we should consider backporting it to unblock those customers.

steveteuber commented:
@dmage I squash the commits.

adambkaplan commented:
/retitle Bug 1756475: Add flag 'DisableRedirect', decide to route all data through the registry

adambkaplan commented:
@steveteuber needs an e2e test for us to backport. 

You can follow the patterns in the `configuration_test.go` to verify that the environment variable is set properly.

https://github.com/openshift/cluster-image-registry-operator/blob/master/test/e2e/configuration_test.go#L508-L566

adambkaplan commented:
Backport BZ filed as well: https://bugzilla.redhat.com/show_bug.cgi?id=1756478

/cherry-pick release-4.2

steveteuber commented:
@adambkaplan I've added the e2e test `TestDisableRedirect`:

```
$ WHAT="TestDisableRedirect" make test-e2e
./hack/test-go.sh -count 1 -timeout 30m -v${WHAT:+ -run="$WHAT"} ./test/e2e/
=== RUN   TestDisableRedirect
--- PASS: TestDisableRedirect (38.25s)
    imageregistry.go:179: creating the image registry resource...
    imageregistry.go:184: starting the operator...
    imageregistry.go:248: waiting for the operator to deploy the registry...
    imageregistry.go:228: waiting for the registry: available (<unset>), progressing (<unset>), Degraded (<unset>), removed (<unset>)
    imageregistry.go:228: waiting for the registry: available (<unset>), progressing (<unset>), Degraded (<unset>), removed (<unset>)
    imageregistry.go:228: waiting for the registry: available (False), progressing (True), Degraded (False), removed (False)
    imageregistry.go:228: waiting for the registry: available (False), progressing (True), Degraded (False), removed (False)
    imageregistry.go:228: waiting for the registry: available (False), progressing (True), Degraded (False), removed (False)
    imageregistry.go:228: waiting for the registry: available (True), progressing (False), Degraded (False), removed (False)
    imageregistry.go:262: the image registry resource reports that the registry is deployed and available
    imageregistry.go:156: uninstalling the image registry...
    imageregistry.go:112: waiting for the registry to be removed: available (False), progressing (False), Degraded (False), removed (True)
    imageregistry.go:160: stopping the operator...
    imageregistry.go:164: deleting the image registry resource...
PASS
ok      github.com/openshift/cluster-image-registry-operator/test/e2e   44.697s
```
